scripts/checks.sh

```diff
@@ -403,21 +403,41 @@ function check_timesync()
 	fi
 }
 
-function check_container_engine()
-{
+function check_container_engine() {
+	local tests=(
+		test_container_engine_running_now
+		test_container_engine_restarts
+		test_container_engine_conn_refused
+	)
+	run_tests "${FUNCNAME[0]}" "${tests[@]}"
+}
+
+function test_container_engine_running_now() {
 	if (! systemctl is-active ${ENG} > /dev/null); then
-		log_status "${BAD}" "${FUNCNAME[0]}" "Container engine ${ENG} is NOT running"
-	else
-		local -i engine_restarts
-		engine_restarts=$(systemctl show -p NRestarts ${ENG} | awk -F= '{print $2}')
-		if (( engine_restarts > 0 )); then
-			local start_timestamp
-			start_timestamp=$(systemctl show -p ExecMainStartTimestamp ${ENG} | awk -F= '{print $2}')
-			log_status "${BAD}" "${FUNCNAME[0]}" "Container engine ${ENG} is up, but has ${engine_restarts} \
-unclean restarts and may be crashlooping (most recent start time: ${start_timestamp})"
-		else
-			log_status "${GOOD}" "${FUNCNAME[0]}" "Container engine ${ENG} is running and has not restarted uncleanly"
-		fi
+		echo "${FUNCNAME[0]}" "Container engine ${ENG} is NOT running"
+	fi
+}
+
+function test_container_engine_restarts() {
+	local -i engine_restarts
+	engine_restarts=$(systemctl show -p NRestarts ${ENG} | awk -F= '{print $2}')
+	if (( engine_restarts > 0 )); then
+		local start_timestamp
+		start_timestamp=$(systemctl show -p ExecMainStartTimestamp ${ENG} | awk -F= '{print $2}')
+		echo "${FUNCNAME[0]}" "Container engine ${ENG} is up, but has ${engine_restarts} \
+restarts and may be crashlooping (most recent start time: ${start_timestamp})"
+	fi
+}
+
+function test_container_engine_conn_refused() {
+	local -i refused_conn_count
+	# This catches three cases:
+	# - supervisor complaint: "Error listening to events: Error: connect ECONNREFUSED /var/run/balena-engine.sock"
+	# - supervisor complaint: "Cannot connect to the balenaEngine daemon at unix:///var/run/balena-engine.sock. Is the balenaEngine daemon running?"
+	# - balena engine complaint: "Error while dialing dial unix /var/run/balena-engine/containerd/balena-engine-containerd.sock: connect: connection refused"
+	refused_conn_count=$(journalctl -a | grep -Ei 'refuse|cannot connect' | grep -Ec 'balena.*sock')
```

Comment:
I think this can be an active test, rather than a log-reading test. Maybe something as simple as a `$ENG ps`? If the engine is not running, the CLI will throw the same error and exit 1.

Reply:
Happy to add that, but is it worth keeping the log-reading test?  I'm thinking by analogy with the tests for engine and service restarts: not just whether there's a problem now, but whether there has been a problem in the past.

Reply:
I'm of two minds here -- on one hand, I see there might be value here (though it might be strictly _diagnostic_ value). On the other hand, log parsing tasks are notoriously brittle and perhaps not worth the effort in this case (a limited log size means this test cannot be dispositive, regardless).

If we consider the cases we can expect:

1. Container engine is up, but containerd is dead & not recovering. A realtime test will catch this.
1. Container engine is up, containerd is up as well but was unstable in the past. We've successfully recovered, interesting to balena but perhaps not to end users.

What do you think? I realize we track unexpected engine restarts, though i suspect it is more likely that typically indicates either a) device instability or b) unsafe user-app operations.

Reply:
You're right about the shortcomings of this test, and of log parsing in general.  I think this is an attempt on my part to get around our lack of `container_engine_connection_problem++`-sort of counters/metrics, and to get quick insight into how things were in the past.  The more I think about this, the more I'm starting to think that this is a less good fit (though not a complete mis-fit) into the way the checks work now.

I do think that having this info in the checks, even if it's mainly interesting to support (rather than the customer), is no bad thing -- it's better than asking a support agent to remember to run the grep command if they suspect there have been more restarts.  To the extent that we want to make our tools and info available to users so those users can fix problems themselves, I think this fits in: it's a scripted version of something we'd check.

However...I'll take this out and stick with the how-are-things-now approach.  The brittleness does concern me, and there's limited real estate/attention in the checks page.  For now, we can add something to JF about checking for connection problems if warranted.